src/superqt/collapsible/_collapsible.py

```diff
@@ -0,0 +1,114 @@
+
+
+class QCollapsible(QFrame):
+    """
+    A collapsible widget to hide and unhide child widgets.
+
+    Based on https://stackoverflow.com/a/68141638
+    """
+
+    _EXPANDED = "▼  "
+    _COLLAPSED = "▲  "
+
+    def __init__(
+        self, title: str = "", parent: Optional[QWidget] = None, flags=Qt.WindowFlags()
+    ):
+        super().__init__(parent, flags)
+        self._locked = False
+
+        self._toggle_btn = QPushButton(self._EXPANDED + title)
+        self._toggle_btn.setCheckable(True)
+        self._toggle_btn.setChecked(False)
+        self._toggle_btn.setStyleSheet("text-align: left; background: transparent;")
+        self._toggle_btn.clicked.connect(self._toggle)
+
+        # frame layout
+        self.setLayout(QVBoxLayout())
+        self.layout().setAlignment(Qt.AlignTop)
+        self.layout().addWidget(self._toggle_btn)
+
+        # Create animators
+        self._animation = QPropertyAnimation(self)
+        self._animation.setPropertyName(b"maximumHeight")
+        self._animation.setStartValue(0)
+        self.setDuration(300)
+        self.setEasingCurve(QEasingCurve.Type.InOutCubic)
+
+        # default content widget
+        _content = QWidget()
+        _content.setLayout(QVBoxLayout())
+        _content.setMaximumHeight(0)
+        self.setContent(_content)
+
+    def setText(self, text: str):
+        current = self._toggle_btn.text()[: len(self._EXPANDED)]
+        self._toggle_btn.setText(current + text)
+
+    def setContent(self, content: QWidget):
+        self._content = content
+        self.layout().addWidget(self._content)
+        self._animation.setTargetObject(content)
+
+    def content(self) -> QWidget:
+        return self._content
+
+    def setDuration(self, msecs: int):
+        self._animation.setDuration(msecs)
+
+    def setEasingCurve(self, easing: QEasingCurve):
+        self._animation.setEasingCurve(easing)
+
+    def addWidget(self, widget: QWidget):
+        self._content.layout().addWidget(widget)
+
+    def removeWidget(self, widget: QWidget):
+        self._content.layout().removeWidget(widget)
+
+    def expand(self, animate: bool = True):
+        if self._locked:
+            return
+
+        self._toggle_btn.setChecked(True)
+        self._toggle_btn.setText(
+            self._EXPANDED + self._toggle_btn.text()[len(self._COLLAPSED) :]
+        )
+        if animate:
+            self._animate(QAbstractAnimation.Direction.Forward)
+        else:
+            self._content.setMaximumHeight(self._content.sizeHint().height() + 10)
+
+    def collapse(self, animate: bool = True):
+        if self._locked:
+            return
+        self._toggle_btn.setChecked(False)
+        self._toggle_btn.setText(
+            self._COLLAPSED + self._toggle_btn.text()[len(self._EXPANDED) :]
+        )
+        if animate:
+            self._animate(QAbstractAnimation.Direction.Backward)
+        else:
+            self._content.setMaximumHeight(0)
```

Comment:
the only downside here is that there's now a lot of code duplication, these are essentially the same methods, with a couple very minor things changed... curious what your thoughts are on this versus just renaming `_animate` ... which really was a very internal thing anyway?

Reply:
@tlambert03 ok. I just merged things into one function to avoid duplication. This means 2 more if statement to execute per expand or collapse but it is probably just a couple of ms extra :) All looks good now

I also removed the top margine from the default content widget. I think it was too far from the title of the collapsible section.

Reply:
> just a couple of ms extra

probably nanoseconds!
```py
In [1]: def f():
   ...:     return
   ...:

In [2]: def f2():
   ...:     if True:
   ...:         pass
   ...:     if False:
   ...:         pass
   ...:     if True:
   ...:         pass
   ...:     return
   ...:

In [3]: %timeit f()
51.6 ns ± 1.67 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)

In [4]: %timeit f2()
56 ns ± 2.08 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)
```

well worth the code dedup :)

> I also removed the top margine from the default content widget. I think it was too far from the title of the collapsible section.

👍

Reply:
Wow didn't know it is this performant.